### ISSUE TITLE
Enable RichTextEffects being applied to characters only once displayed

### ISSUE THREAD

**Issue by IntangibleMatter** [godotengine/godot-proposals#7267]
### Describe the project you are working on

An rpg with a lot of text with effects and styles, and dialogue that displays at variable rates.

### Describe the problem or limitation you are having in your project

Because of how Godot's RichTextLabel is designed, it is impossible to do all three of the following at the same time:

- Have a RichTextLabel's text be laid out properly ahead of time
- Have every new character appear with a text effect animating them in
- Have text display at a variable rate

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose adding a flag to the RichTextLabel (or RichTextEffect) that enables the time which its animation is based on starting from when the character is visible, rather than the initialization of the RichTextLabel. This would allow people to easily make an effect that animates each character on their display, allowing for smoother display of characters.

Alternately, a new type of BBCode could be added, called something to the effect of `RichTextEffectTransition`. This type of effect could be wrapped around an entire string, and applied once to each character when they first become visible.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

There are a few ways it could work, and I'm uncertain on what method would be best to work with the existing code while still having good performance on large bodies of text. But approximately the input of the RichTextLabel would be something to the effect of

```txt
[fadein]Well, would you look at [wave]who just came in??[/wave] Glad to see you back, dude![/fadein]
```
With `fadein` being a RichTextEffect transition or applied at display time.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

As far as I know this would be a nightmare to work around, you would have to do the following:

1. Calculate the positions of newlines in the RichTextLabel
2. Insert those newlines into the raw text (could hypothetically be done ahead of time, but would be a nightmare with translations)
3. Add each character with their effect to the RichTextLabel one at a time for it to display properly

I don't know if Godot even has the functionality to tell you where newlines are in a RichTextLabel built in.

### Is there a reason why this should be core and not an add-on in the asset library?

For this to be performant and not a nightmare to implement, it couldn't be done as an add-on.

**Comment by IntangibleMatter** [comment]
I found this discussion #5880 which mentions adding an `_on_visible()` callback to RichTextEffect, which would also be a viable option, and possible easier to implement.

**Comment by IntangibleMatter** [comment]
Superseded by #7270
